### DISCUSSION TITLE
How to interpret stumpy.motifs?

### DISCUSSION THREAD

**Discussion by SaVoAMP** [TDAmeritrade/stumpy#498] [Q&A] [answered]
Hey, me again!

I have trouble understanding stumpy.motifs.
I had understood the API to find a desired number of motifs (`max_motifs`) after calling `stumpy.motifs` as well as their nearest neighbors (`max_matches`). Or can I just search different motifs in `stumpy.motifs` and need `stumpy.macth` to get their nearest neighbors as well?

I have written the following lines of code:
```
mp_pre = stumpy.gpu_stump(recording['pre_mag_magnitude'], sub_len)
motifs = stumpy.motifs(T=recording['pre_mag_magnitude'], P=mp_pre[:, 0], max_matches=10, max_motifs=3)
```
As a result I get a tuple with two ndarrays of size (1,2) - i.e. the motif_distances and motif_indices.
How can I interpret this now? Why do the arrays only have shape (1,2) and not (3,2) when I am looking for 3 different motifs? Concretely I receive `motifs[0][0, 0] = 0` (self match) and `motifs[0][0, 1] = 8.06108` (probably the distance to the nearest neighbor?) with the corresponding indices `motifs[1][0, 0] = 9823` and `motifs[1][0, 1] = 9119`.  Wouldn't I have to have two more rows each for the other two motif pairs here? 
And where is my `max_matches` information? I thought that by setting `max_matches=10` I would also automatically get the 10 nearest neighbors for each of the subjects as with `stumpy.match`?

**Comment by seanlaw** [comment, marked as answer]
@SaVoAMP Good question. It sounds like we need to add some more details to the documentation to make these things clearer. So, in an attempt to conserve memory/space, we opted to only return essential information to the user that satisfies several key criterion that are defined by the input parameters. When you set `max_matches=10`, this is the upper limit as to how many matches are allowed to be returned for each "candidate motif". As you can imagine, for each "candidate motif", it is possible for it to have many, many matches (i.e., beyond 10) OR it can have no matches (i.e., when the distance between the candidate and the matches is greater than `max_distance`). Similarly, when you set `max_motifs=3`, this is the upper limit as to how many motifs are allowed to be returned. Note that a "candidate motif" must have at least one nearest neighbor that is within a distance that is equal to or smaller than `cutoff`. 

So, in theory, if everything works out nicely and you have great candidate motifs (that aren't disqualified) that then have an abundance of matches (that aren't disqualified), then you should expect to get a `motif_distances` of shape `(3, 10)`. However, since you are getting something of size `(1, 2)` implies that only one candidate motif was found (within some `cutoff` criteria) and it had two matches (that had a distance below `max_distance`). All other candidate motifs had a distance to a nearest neighbor that was larger than the default `cutoff` distance AND the other matches have a distance to the candidate motif that is larger than `max_distance`. 

Of course, you can set the `cutoff` to something large like `np.inf` and also set the `max_distance` to something large like `np.inf` and then you'll guarantee that the arrays will come back with shape `(3, 10)` (or, in general, shape `(max_motifs, max_matches)`). However, the key here is that not all matches are created equal (i.e., the quality of the matches will differ by their distance to the candidate motif) and not all motifs are created equal (i.e., you can have motifs that have a one nearest neighbor that is quite far away). So, the default `cutoff` and `max_distance` apply some reasonable controls to ensure that the identified motifs and their matches are significant (namely, less than 2 standard deviations from the mean distance).

Now that I think about it, perhaps it would make more sense to always make sure that the returned arrays have shape `(max_motifs, max_matches)` but then "blank out" the parts where the `cutoff`/`max_distance` constraints were not satisfied.

Let me know if that makes sense. I'd be curious to get your feedback so that we can improve this.

**Reply by SaVoAMP** [reply]
Ahh yes, that makes sense, thank you!

> Now that I think about it, perhaps it would make more sense to always make sure that the returned arrays have shape (max_motifs, max_matches) but then "blank out" the parts where the cutoff/max_distance constraints were not satisfied.

That sounds very reasonable to me! This is because I know from my already labeled data set how many motifs I am looking for and how many matches I would like to have. I would try to make it a bit clearer that the output array has shape `(max_motifs, max_matches)`. 
But it would be really great if on the one hand it would stay as it already is and on the other hand you could decide yourself how many motifs and matches you want to have, no matter for which `cutoff / max_distance` values (so exactly as you suggested)!

**Reply by seanlaw** [reply]
In order to cause the least disruption to existing users (i.e., without changing the array outputs), what if I added the following to the docstring:

```
    Note that, in the best case scenario, the returned arrays would have shape
    `(max_motifs, max_matches)` and contain all finite values. However, in reality,
    many conditions (see below) need to be satisfied in order for this to be true. Any
    truncation in the number of rows (i.e., motifs)  may be the result of insufficient
    candidate motifs with matches greater than or equal to `min_neighbors` or that the
    matrix profile value for the candidate motif was larger than `cutoff`. Similarly,
    any truncationin in the number of columns (i.e., matches) may be the result of
    insufficient matches being found with distances (to their corresponding candidate
    motif) that are equal to or less than `max_distance`. Only motifs and matches that
    satisfy all of these constraints will be returned.

    If you must return a shape of `(max_motifs, max_matches)`, then you may consider
    specifying a smaller `min_neighors`, a larger `max_distance`, and/or a larger
    `cutoff`. For example, while it is not advised, setting `min_neighbors=1`, `max_distance=np.inf`, and
    `cutoff=np.inf` will ensure that the shape of the output arrays will be
    `(max_motifs, max_matches)`. However, given the lack of constraints, the quality
    of each motif and the quality of each match may be drastically different. Setting
    appropriate conditions will help ensure appropriately constrained results that may
    be easier to interpret.
```

**Reply by SaVoAMP** [reply]
> In order to cause the least disruption to existing users (i.e., without changing the array outputs), what if I added the following to the docstring:

yes, I think the proposal is great! Now everything is clearly understandable!

**Reply by seanlaw** [reply]
Done https://stumpy.readthedocs.io/en/latest/api.html#motifs

Thank you for your feedback!

**Reply by SaVoAMP** [reply]
I tried to execute stumpy.match with the parameters `cutoff = np.inf` and ´max_distance = np.inf´ but it did not work.
There  was no error message but the program seems to get stuck in a loop somewhere.

Strangely setting ´cutoff = np.inf´ only, gets me the right number of motifs but using ´max_distance = np.inf´ gets the program stuck.

I'm only able to get the desired number of matches using ´stumpy.match()´ and setting the same parameter ´max_distance = np.inf´ seems to work fine in this case.

**Reply by seanlaw** [reply]
> I tried to execute stumpy.match with the parameters `cutoff = np.inf` and ´max_distance = np.inf´ but it did not work.
> 
> There  was no error message but the program seems to get stuck in a loop somewhere.
> 
> 
> 
> Strangely setting ´cutoff = np.inf´ only, gets me the right number of motifs but using ´max_distance = np.inf´ gets the program stuck.
> 
> 
> 
> I'm only able to get the desired number of matches using ´stumpy.match()´ and setting the same parameter ´max_distance = np.inf´ seems to work fine in this case.

There is a small bug that was just fixed when `max_distance=np.inf` (see Issue #501). Can you please try setting it to an absurdly large finite number like `max_distance=1e20` instead?

**Reply by SaVoAMP** [reply]
Ahh, thank you!
I updated stumpy to the latest version and it works now.

**Reply by seanlaw** [reply]
Glad to hear that. Happy holidays, @SaVoAMP!

**Reply by SaVoAMP** [reply]
Thank you, I wish you happy holidays too!
